scio-smb/src/main/scala/com/spotify/scio/smb/syntax/SortMergeBucketSCollectionSyntax.scala

```diff
@@ -49,16 +52,30 @@ final class SortedBucketSCollection[T](private val self: SCollection[T]) {
    *   contains information about key function, bucket and shard size, etc.
    */
   @experimental
-  def saveAsSortedBucket(write: SortedBucketIO.Write[_, _, T]): ClosedTap[Nothing] = {
+  def saveAsSortedBucket(
+    write: SortedBucketIO.Write[_, _, T]
+  )(implicit coder: Coder[T]): ClosedTap[T] = {
     if (self.context.isTest) {
-      val testOutput = TestDataManager.getOutput(self.context.testId.get)
-      testOutput[T](SortedBucketIOUtil.testId(write))(self)
+      TestDataManager.getOutput(self.context.testId.get)(SortedBucketIOUtil.testId(write))
+      ClosedTap(TapOf[T].saveForTest(self))
     } else {
-      self.applyInternal(write)
-    }
+      val writeResult = self.applyInternal(write).expand()
 
-    // @Todo: Implement taps for metadata/bucket elements
-    ClosedTap[Nothing](EmptyTap)
+      val bucketFiles = self.context
+        .wrap(
+          writeResult
+            .get(new TupleTag("WrittenFiles"))
+            .asInstanceOf[PCollection[KV[BucketShardId, ResourceId]]]
+        )
+        .materialize
+
+      val fileOps = write.getFileOperations
+
+      ClosedTap(
+        bucketFiles.underlying
+          .flatMap(kv => fileOps.iterator(kv.getValue).asScala)
```

Comment:
We could keep the `BucketShardId` so the tap returns a mapping from Bucket ID --> records in that bucket - although we'd have to also find a way to implement this in test context (probably by making another `SortedBucketIO.Write#getBucketMetadata` public too and mapping the bucketing function over each elements)